consensus/istanbul/celo-ibft-spec.md

```diff
@@ -0,0 +1,488 @@
+
+Participants are able to broadcast messages to all other participants. In the
+case that a participant is off-line or somehow inaccessible they will not
+receive broadcast messages and there is no mechanism for these messages to be
+re-sent.
+
+We refer to a consensus instance to mean consensus for a specific height,
+consensus instances are independent (they have no shared state), that is
+messages or values from the previous height have no bearing on the outcome of
+the current height
+
+We assume the existence of an application that feeds values to be agreed upon
+to the consensus instance and also receives agreed upon values from the
+consensus. The application also provides application specific implementations
+for certain functions that the consensus algorithm relies upon, such as a
+function that determines the validity of a value. 
+
+### Algorithm
+
+See supporting [functions](#Functions) and [notation](#Appendix-1-Notation).
+
+Note all state modifications are contained in the following code block,
+supporting functions cannot modify state, they all operate without side
+effects.
+
+Variable names                  |Instance state                          
+--------------------------------|----------------------------------------
+`H - height`                    |`Hc - current height`                                      
+`R - round          `           |`Rc - current round`                    
+`V - value`                     |`Rd - desired round`                    
+`T - Message type`              |`Vc - currently proposed value`         
+`RCC - round change certificate`|`PCc - current prepared certificate`    
+`PC - prepared certificate`     |`Sc - current participant state`        
+`S - participant state`         |
+`M or N - message sets`         |
+
+```
+
+// Upon receiving this event the algorithm transitions to the next height the
+// final committed event signifies that the application has accepted the agreed
+// upon value for the current height.
+upon: <FinalCommittedEvent>
+  Hc ← Hc+1
+  Rc ← 0
+  Rd ← 0
+  Sc ← AcceptRequest
+  Vc ← nil
+  schedule onRoundChangeTimeout(Hc, 0) after roundChangeTimeout(0)
+
+// A request event is a request to reach agreement on the provided value, the
+// request event is sent by the application, if this parcicipant is the proposer
+// it will propose that value by sending a preprepared message.
+upon: <RequestEvent, Hc, V> && Sc = AcceptRequest
+  if Rc = 0 && isProposer(Hc, Rd) {
+    bc(<Preprepare, Hc, 0, V, nil>)
+  }
+
+// When receiving a preprepare from a proposer participants will vote for the
+// value (if valid) by sending a prepare message.
+upon: <Preprepare, Hc, Rd, V, RCC> from proposer(Hc, Rd) && Sc = AcceptRequest
+  if (Rd > 0 && validRCC(V, RCC)) || (Rd = 0 && RCC = nil)  {
+    Rc ← Rd
+    Vc ← V
+    Sc ← Preprepared
+    bc(<Prepare, Hc, Rd, Vc>)
+  }
+
+// When a participant sees at least 2f+1 prepare or commit messages for a value
```

Comment:
what's not clear from the spec, is that message set contains a single message per participant. So having the prepare and the commit from the same particpant doesn't count as `|M| = 2`

Reply:
Good point, I thought I had addressed this in the system model section, but I had not.

Reply:
I'd actually define what message set M is. There's also a N set that i haven't seen used. 

In any case. If we do this by the operaor. I wouldn't use || as that since people understand the symbol as size in a set. So i wouldn't override that notion. Instead use another function

Reply:
So M is just a variable that represents a set of messages, what is in M is defined by whatever statement is on the right of the assignment operator, N is used in the case that we need to represent more than one set in a single scope.

I've updated the appendix to make this more clear, let me know if it is not.

https://github.com/celo-org/celo-blockchain/blob/ec196aa1e40ef304d49142d172265ea151735444/consensus/istanbul/celo-ibft-spec.md?plain=1#L311-L312